### PR TITLE
Add Mac Build JDK From SBOM script

### LINKED ISSUE #3633: Add Mac support for building from SBOM script
The equivalent of [https://github.com/adoptium/temurin-build/blob/master/tooling/linux_repro_build_compare.sh](https://github.com/adoptium/temurin-build/blob/master/tooling/linux_repro_build_compare.sh)  
needs adding for  Mac

Seperated from #3426

### PR DESCRIPTION
**Pull request by steelhead31**
Fixes #3633 

Has been tested extensively as a standalone script: sample output:

```
Comparing src_jdk/Contents/Home with tar_jdk/Contents/Home ... output to file: repro_diff.out
Number of differences:      154
ReproduciblePercent = 99.85 %

```

### CODE REVIEW

**Comment by steelhead31** [comment]
This script is the mac equivalent of the windows script :)

File: tooling/reproducible/macos_repro_build_compare.sh [hunk 1]
```diff
@@ -0,0 +1,673 @@
...
+# Read Parameters
+SBOM_URL="$1"
+TARBALL_URL="$2"
+
+# Constants Required By This Script
+# These Values Should Be Updated To Reflect The Build Environment
+# The Defaults Below Are Suitable For An Adoptium Windows Build Environment
+# Which Has Been Created Via The Ansible Infrastructure Playbooks
+
+WORK_DIR=/Users/jenkins/sfryer/comp-jdk-build
```
Comment (andrew-m-leonard) [inline tooling/reproducible/macos_repro_build_compare.sh:46]: I suspect WORK_DIR needs to be a argument or sub-folder?

File: tooling/reproducible/macos_repro_build_compare.sh [hunk 2]
```diff
@@ -0,0 +1,673 @@
...
+}
+
+Compare_JDK() {
+  echo "Comparing JDKs"
+  echo ""
+  mkdir "$WORK_DIR/compare"
+  cp "$WORK_DIR/src_jdk_dist.tar.gz" "$WORK_DIR/compare"
+  cp "$WORK_DIR/built_jdk.tar.gz" "$WORK_DIR/compare"
+
+  # Get The Current Versions Of The Reproducible Build Scripts
```
Comment (andrew-m-leonard) [inline tooling/reproducible/macos_repro_build_compare.sh:572]: Possibly ought to integrate some of Sophia's latest updates to the Compare_JDK() function - for Windows : https://github.com/adoptium/temurin-build/commit/ad4a1c3ae345a6f9b721b5fbea9ebfc0f1005a83 ?

File: tooling/reproducible/macos_repro_build_compare.sh (hunk 1 above)
Comment (steelhead31) [reply tooling/reproducible/macos_repro_build_compare.sh]: Fixed.

File: tooling/reproducible/macos_repro_build_compare.sh (hunk 2 above)
Comment (steelhead31) [reply tooling/reproducible/macos_repro_build_compare.sh:578]: Done, aligned those elements I think required it.

File: tooling/reproducible/macos_repro_build_compare.sh [hunk 3]
```diff
@@ -0,0 +1,686 @@
...
+  #Unzip And Rename The Target JDK
+  echo "Unzip Target"
+  tar xvfz reproJDK.tar.gz
+  original_directory_name=$(find . -maxdepth 1 -type d | grep -v src_jdk | tail -1)
+  mv "$original_directory_name" tar_jdk
+
+  # These Two Files Are Generate Classes And Should Be Removed Prior To Running The Comparison
+  # jdk/bin/server/classes.jsa & jdk/bin/server/classes_nocoops.jsa
+
+  if [ -f "$WORK_DIR/compare/src_jdk/bin/server/classes.jsa" ] ; then
```
Comment (andrew-m-leonard) [inline tooling/reproducible/macos_repro_build_compare.sh:602]: These are now removed by https://github.com/adoptium/temurin-build/blob/f4f23184b8102e1d069d2b9def13b5a6d0aa1b04/tooling/reproducible/repro_common.sh#L219 just needs a Darwin check and should use mac /Contents/Home/.. paths...?

Comment (steelhead31) [reply tooling/reproducible/macos_repro_build_compare.sh:602]: Ah, good spot, I'll fix that too.

Comment (steelhead31) [reply tooling/reproducible/macos_repro_build_compare.sh]: Ah JDK\_DIR is already passed in with Contents/Home appended , so the above should work.

File: tooling/reproducible/macos_repro_build_compare.sh [hunk 4]
```diff
@@ -0,0 +1,685 @@
...
+fi
+
+# Read Parameters
+SBOM_URL="$1"
+TARBALL_URL="$2"
+REPORT_DIR="$3"
+
+# Constants Required By This Script
+# These Values Should Be Updated To Reflect The Build Environment
+# The Defaults Below Are Suitable For An Adoptium Windows Build Environment
```
Comment (karianna) [inline tooling/reproducible/macos_repro_build_compare.sh:46]:
```suggestion
# The Defaults Below Are Suitable For An Adoptium Mac OS X Build Environment
```

File: tooling/reproducible/macos_repro_build_compare.sh [hunk 5]
```diff
@@ -0,0 +1,685 @@
...
+MAC_COMPILER_BASE=/Applications
+MAC_COMPILER_APP_PREFIX=Xcode
+MAC_SDK_LOCATION=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk
+
+# These 3 variables dictate which parameters passed to makejdk_any_platform.sh are config arguments, build arguments or should be ignored.
+CONFIG_ARGS=("--disable-warnings-as-errors" "--openjdk-target" "--with-sysroot" "--with-extra-cxxflags='" "--enable-dtrace" "--with-version-opt")
+NOTUSE_ARGS=("--assemble-exploded-image" "--configure-args")
+FINAL_ARG=("--build-variant")
+
+
```
Comment (karianna) [inline tooling/reproducible/macos_repro_build_compare.sh:59]:
```suggestion
```

File: tooling/reproducible/macos_repro_build_compare.sh [hunk 6]
```diff
@@ -0,0 +1,685 @@
...
+      cp "$TARBALL_URL" "$DISTLocalPath"
+    else
+      echo "ERROR - The Supplied Path To The JDK Zipfile Does Not Exist"
+      exit 1
+    fi
+  fi
+}
+
+Install_PreReqs() {
+
```
Comment (karianna) [inline tooling/reproducible/macos_repro_build_compare.sh:146]:
```suggestion
```

File: tooling/reproducible/macos_repro_build_compare.sh [hunk 7]
```diff
@@ -0,0 +1,685 @@
...
+  # Check if the SBOM has the expected structure
+  if [ -z "$sbomContent" ] || [ "$(echo "$sbomContent" | jq -r '.metadata.tools')" == "null" ]; then
+    echo "Invalid SBOM format. Unable to extract Data."
+    exit 1
+  else
+    echo "SBOM Is Structurally Sound.. Extracting Values:"
+    echo ""
+  fi
+
+    # Extract All Required Fields From The SBOM Content
```
Comment (karianna) [inline tooling/reproducible/macos_repro_build_compare.sh:177]: whitespace incorrect for this whole block?

**Review by karianna — APPROVED** [review]

File: tooling/reproducible/macos_repro_build_compare.sh [hunk 8]
```diff
@@ -0,0 +1,686 @@
...
+  original_directory_name=$(find . -maxdepth 1 -type d | tail -1)
+  mv "$original_directory_name" src_jdk
+
+  #Unzip And Rename The Target JDK
+  echo "Unzip Target"
+  tar xvfz reproJDK.tar.gz
+  original_directory_name=$(find . -maxdepth 1 -type d | grep -v src_jdk | tail -1)
+  mv "$original_directory_name" tar_jdk
+
+  # These Two Files Are Generate Classes And Should Be Removed Prior To Running The Comparison
```
Comment (andrew-m-leonard) [inline tooling/reproducible/macos_repro_build_compare.sh:599]: The removal of the classes.jsa is done by repro_common.sh now, just need to add for Darwin here: https://github.com/adoptium/temurin-build/blob/298347ac59e6b73b15d99bce58b66a6d44cce940/tooling/reproducible/repro_common.sh#L218

Comment (steelhead31) [reply tooling/reproducible/macos_repro_build_compare.sh]: Fixed.

File: tooling/reproducible/macos_repro_build_compare.sh [hunk 9]
```diff
@@ -0,0 +1,666 @@
...
+  
+
+  # Get The Current Versions Of The Reproducible Build Scripts
+  wget -O "$WORK_DIR/compare/repro_common.sh" "https://raw.githubusercontent.com/adoptium/temurin-build/master/tooling/reproducible/repro_common.sh"
+  wget -O "$WORK_DIR/compare/repro_compare.sh" "https://raw.githubusercontent.com/adoptium/temurin-build/master/tooling/reproducible/repro_compare.sh"
+  wget -O "$WORK_DIR/compare/repro_process.sh" "https://raw.githubusercontent.com/adoptium/temurin-build/master/tooling/reproducible/repro_process.sh"
+
+  # Set Permissions
+  chmod +x "$WORK_DIR/compare/"*sh
+  cd "$WORK_DIR/compare"
```
Comment (sophia-guo) [inline tooling/reproducible/macos_repro_build_compare.sh:585]: Those files are part of  tool/reproducible. No need to require. Especially it's hard code to temurin-build/master, which won't take developer's PRs.

Comment (steelhead31) [reply tooling/reproducible/macos_repro_build_compare.sh:585]: I'll update the script, so it checks for their existence locally, and only download the versions from git, if they are not found for any reason :)

File: tooling/reproducible/repro_common.sh [hunk 10]
```diff
@@ -332,7 +332,7 @@ function removeGeneratedClasses() {
   local JDK_DIR="$1"
   local OS="$2"
 
-  if [[ "$OS" =~ CYGWIN* ]]; then
+  if [[ "$OS" =~ CYGWIN* ]] || [[ "$OS" =~ Darwin* ]]; then
     rm -rf "$JDK_DIR/bin/server/classes.jsa"
```
Comment (andrew-m-leonard) [inline tooling/reproducible/repro_common.sh:336]: I suspect the paths will need Contents/Home for Mac maybe?

Comment (steelhead31) [reply tooling/reproducible/repro_common.sh:336]: I have my script include them in the JDK\_DIR being passed, which means this script stays generic, but the windows & ,mac scripts pass in the correct dir :)

File: tooling/reproducible/macos_repro_build_compare.sh (hunk 9 above)
Comment (steelhead31) [reply tooling/reproducible/macos_repro_build_compare.sh:598]: This change is in the current version of the script.

**Review by andrew-m-leonard — APPROVED** [review]
looks good

**Pull request merged**